[IMP] inventory: sell stock virtual warehouse (#7415)
odoo/documentation: content/applications/inventory_and_mrp/inventory/routes/concepts/stock_warehouses.rst

> In Odoo, sometimes products included in a single sales order might take stock from two (or more)
> warehouses. In Odoo, pulling products from multiple warehouses to satisfy sales demands can be done
> using *virtual locations*.

Reviewer: *In Odoo* intro twice in a row. Quick fix ⬇️ 
```suggestion
Sometimes products included in a single sales order might take stock from two (or more)
warehouses; in Odoo, pulling products from multiple warehouses to satisfy sales demands can be done
using *virtual locations*.
```

warehouse's **stock location** (e.g. `VWH/Stock`) from the drop-down menu, and click
:guilabel:`Save`.

.. important::
   To select the virtual warehouse's stock location in the :guilabel:`Parent Location` drop-down
   menu, the parent warehouse stock location (e.g. `VWH/Stock`) **must**  have its
   :guilabel:`Location Type` set to :guilabel:`Internal Location`.

.. image:: stock_warehouses/configure-physical-wh.png
   :align: center
   :alt: Set the child warehouse's *Parent Location* to the virtual warehouse.

Repeat the preceding steps to configure two or more child warehouses.

Once complete, the virtual, parent warehouse (e.g. `VWH/Stock`) fulfills orders using stock from
child warehouses (e.g. `WHA` and `WHB`), if there is insufficient stock in any one location.

Set virtual stock location as 'view'
====================================

.. note::
   (Optional) Set the virtual stock location's :guilabel:`Location Type` to :guilabel:`View`, as it
   is a non-existent location used to group various physical warehouses together.

.. tip::
   To view the quantity of a product across **all** linked child warehouses, go to the product form

>    and click the :guilabel:`On Hand` smart button.

Author: Does this make more sense now, @hojo-odoo ? 
This is to address this request from a BSA in BE:
![image](https://github.com/odoo/documentation/assets/23611711/e371bc99-131e-4708-abc4-d7eb4e9aaddd)

>    click the :guilabel:`On Hand` smart button.

Reviewer 2: This just doesn't make sense to me. You say that the location needs to be set to "View" to create a hierarchical structure, but then I don't see any mention of that structure after the opening sentence. The rest just talks about viewing the on-hand quantity across all warehouses.

>    Change the virtual stock location to :guilabel:`View` to create a hierarchical structure for a
>    warehouse. Doing so displays the total on-hand quantity of the product across **all** linked
>    child warehouses.
> 
>    To view the quantity of a product across all linked child warehouses, go to the product form and
>    click the :guilabel:`On Hand` smart button.

Author: ```suggestion
   Optional: Set the virtual stock location's :guilabel:`Location Type` to :guilabel:`View` since it is a non-existent location used to group various physical warehouses together.
   
.. tip::

   To view the quantity of a product across **all** linked child warehouses, go to the product form and
   click the :guilabel:`On Hand` smart button.
```

   To do so, go to :menuselection:`Inventory --> Configuration --> Settings`, scroll down to the
   :guilabel:`Warehouse` section, and enable the :guilabel:`Storage Locations` and
   :guilabel:`Multi-Step Routes` options. Then, :guilabel:`Save` the changes to finish.

.. _inventory/routes/virtual-wh:

Create virtual parent location
==============================

Before creating any virtual stock locations, create a new warehouse that acts as a *virtual*
warehouse, the *parent* location of other physical warehouses.

.. spoiler:: Why a "virtual" warehouse?

   Virtual warehouses are great for companies with multiple physical warehouses. This is because a
   situation might arise when one warehouse runs out of stock of a particular product, but another
   warehouse still has stock on-hand. In this case, stock from these two (or more) warehouses could
   be used to fulfill a single sales order.

   The "virtual" warehouse acts as a single aggregator of all the inventory stored in a company's
   physical warehouses, and is used (for traceability purposes) to create a hierarchy of locations
   in Odoo.

> To create a new warehouse, go to :menuselection:`Inventory --> Configuration --> Warehouses`, and
> click :guilabel:`Create`. From here, the warehouse :guilabel:`Name` and :guilabel:`Short Name` can
> be changed, and other warehouse details can be changed under the :guilabel:`Warehouse Configuration`
> tab.

Reviewer 3: ```suggestion
To create a new warehouse, go to :menuselection:`Inventory app --> Configuration --> Warehouses`,
and click :guilabel:`Create`. From here, the warehouse :guilabel:`Name` and :guilabel:`Short Name`
can be changed, and other warehouse details can be changed under the :guilabel:`Warehouse
Configuration` tab.
```

Lastly, click :guilabel:`Save` to finish creating a *regular* warehouse. Continue following the
steps below to finish configuring the virtual parent warehouse.

.. image:: stock_warehouses/stock-warehouses-create-warehouse.png
   :align: center
   :alt: New warehouse form.

.. seealso::
   - :doc:`Warehouse configurations <../../management/warehouses/warehouses_locations>`
   - :ref:`Incoming and outgoing shipments <inventory/receipts_delivery_one_step/wh>`
   - :doc:`Resupply from another warehouse <../../management/warehouses/resupply_warehouses>`

.. _inventory/routes/child-wh:

Create child warehouses
=======================

> Next, create at least two *child* warehouses to link to the virtual warehouse.

Reviewer 3: ```suggestion
Create at least two *child* warehouses to link to the virtual warehouse.
```
just so these sections can standalone, I'd suggest removing the "Next" from this sentence.

.. important::

>    In order to take stock from multiple warehouses to fulfill a sales order, there need to be at

Reviewer 3: ```suggestion
   In order to take stock from multiple warehouses to fulfill a sales order, there needs to be at
```

Change the :guilabel:`Parent Location` field from :guilabel:`Physical Locations` to the virtual
warehouse's **stock location** (e.g. `VWH/Stock`) from the drop-down menu, and click
:guilabel:`Save`.

.. important::
   To select the virtual warehouse's stock location in the :guilabel:`Parent Location` drop-down
   menu, the parent warehouse stock location (e.g. `VWH/Stock`) **must**  have its
   :guilabel:`Location Type` set to :guilabel:`Internal Location`.

.. image:: stock_warehouses/configure-physical-wh.png
   :align: center
   :alt: Set the child warehouse's *Parent Location* to the virtual warehouse.

Repeat the preceding steps to configure two or more child warehouses.

Once complete, the virtual, parent warehouse (e.g. `VWH/Stock`) fulfills orders using stock from
child warehouses (e.g. `WHA` and `WHB`), if there is insufficient stock in any one location.

Set virtual stock location as 'view'
====================================

Set the virtual stock location's :guilabel:`Location Type` to :guilabel:`View`, as it is a
non-existent location used to group various physical warehouses together.

> To do that, navigate to :menuselection:`Inventory --> Configuration --> Locations`.

Reviewer 3: ```suggestion
To do that, navigate to :menuselection:`Inventory app --> Configuration --> Locations`.
```

Click the virtual warehouse's stock location (e.g. `VWH/Stock`) that was :ref:`previously created
<inventory/routes/virtual-wh>`, from the :guilabel:`Locations` list.

On the location form, under the :guilabel:`Additional Information` heading, set the
:guilabel:`Location Type` to :guilabel:`View`. :guilabel:`Save` the changes.

.. image:: stock_warehouses/set-location-type-view.png
   :align: center
   :alt: Warehouse location types in location creation screen.

.. tip::
   To view the total quantity across **all** linked child warehouses, go to the product form and
   click the :guilabel:`On Hand` smart button.

   .. image:: stock_warehouses/on-hand.png
      :align: center
      :alt: Display stock across all linked warehouses.

> Example: Sell products from a virtual warehouse

Reviewer 3: ```suggestion
Example: sell products from a virtual warehouse
```
i know that ZST tends to prefer lowercase following a colon

===============================================

To sell products from multiple warehouses using a virtual parent location, the database must have at

> least **two** warehouses configured — with at least **one** product with quantity on-hand in each

Reviewer 3: ```suggestion
least **two** warehouses configured — with at least **one** product, with quantity on-hand in each
```

warehouse, respectively.

.. example::
   The following product, `Toy soldier`, is available at each location with the quantities:

   - `WHA/Stock` : 1
   - `WHB/Stock` : 2
   - Warehouses `WHA` and `WHB` are child warehouses of the virtual warehouse `VWH`.

Create a quotation for the product by navigating to the :menuselection:`Sales` app and clicking
:guilabel:`Create`. On the quote, add a :guilabel:`Customer`, and click :guilabel:`Add a product` to
add the two products stored in the two warehouses.

Then, click the :guilabel:`Other Info` tab on the sales order form. Under the :guilabel:`Delivery`
section, change the :guilabel:`Warehouse` field value to the virtual warehouse that was
:ref:`previously created <inventory/routes/virtual-wh>`. Next, :guilabel:`Confirm` the sales order.

.. image:: stock_warehouses/set-virtual-wh.png
   :align: center
   :alt: Set virtual warehouse as the *Warehouse* field in sales order's *Other Info* tab.

> Next, click the :guilabel:`Delivery` smart button. From the warehouse delivery form, confirm that

Reviewer 3: ```suggestion
Then, click the :guilabel:`Delivery` smart button. From the warehouse delivery form, confirm that
```

the :guilabel:`Source Location` value matches the :guilabel:`Warehouse` field value from the sales
order. Both should list the virtual warehouse location.

Finally, on the warehouse delivery form, under the :guilabel:`Detailed Operations` tab, confirm that
the :guilabel:`Locations` in the :guilabel:`From` column for each product match the child locations
that are tied to the virtual parent location.

.. image:: stock_warehouses/delivery-order.png
   :align: center
   :alt: Delivery order with matching source and child locations.

.. important::

>    The :guilabel:`Source Location` on the warehouse delivery form and the :guilabel:`Warehouse`
>    under the :guilabel:`Other Info` tab on the sales order **must** match for products in the sales
>    order to be pulled from different warehouses.

Reviewer 3: ```suggestion
   The :guilabel:`Source Location` on the warehouse delivery form, and the :guilabel:`Warehouse`
   under the :guilabel:`Other Info` tab on the sales order, **must** match for products in the sales
   order to be pulled from different warehouses.
```

  - If the virtual warehouse is not in the :guilabel:`Source Location` field on the warehouse
    delivery form, retry product reservation by:

    - Running the scheduler: turn on :ref:`developer mode <developer-mode>`, and then go to
      :menuselection:`Inventory app --> Operations --> Run Scheduler`.
    - Clicking :guilabel:`Check Availability` on the delivery order.
  - If the virtual warehouse is **not** assigned to the :guilabel:`Warehouse` field on the sales

>     order, then cancel it and create a new sales order with the virtual warehouse set in the

Reviewer 3: ```suggestion
    order, then cancel it, and create a new sales order with the virtual warehouse set in the
```

--> Configuration --> Locations` list.

Remove any filters from the search bar. Then, click the physical warehouse :guilabel:`Location` that
was previously created to be a child location (e.g. `WHA`), and click :guilabel:`Edit`.

Change the :guilabel:`Parent Location` field from :guilabel:`Physical Locations` to the virtual
warehouse's **stock location** (e.g. `VWH/Stock`) from the drop-down menu, and :guilabel:`Save`
changes.

.. important::
   To select the virtual warehouse's stock location in the :guilabel:`Parent Location` drop-down
   menu, the parent warehouse stock location (e.g. `VWH/Stock`) **must**  have its
   :guilabel:`Location Type` set to :guilabel:`Internal Location`.

.. image:: stock_warehouses/configure-physical-wh.png
   :align: center
   :alt: Set the child warehouse's *Parent Location* to the virtual warehouse.

Repeat the preceding steps to configure two or more child warehouses.

Once complete, the child locations of with the same virtual warehouse set as the parent (e.g. `WHA`
and `WHB`) can use products from the others' stock to fulfill a single sales order, if there is

> insufficient stock in any one location.

Author: Addressed this comment in the content:
> Once complete, the virtual, parent warehouse (e.g. `VWH/Stock`) fulfills orders using stock from child warehouses (e.g. `WHA` and `WHB`), if there is insufficient stock in any one location.

Set virtual stock location as 'view'
====================================

.. important::
   Setting the virtual stock location to :guilabel:`View` is necessary for creating a hierarchical

>    structure for a warehouse and aggregate its child locations.

Author: This is the adjustment I made to address your comment!
>    Change the virtual stock location to :guilabel:`View` to create a hierarchical structure for a
   warehouse. Doing so displays the total on-hand quantity of the product across **all** linked
   child warehouses.

>   To view the quantity of a product across all linked child warehouses, go to the product form and
   click the :guilabel:`On Hand` smart button.

.. tip::
   Change the virtual stock location to :guilabel:`View` to see the total on-hand quantity of
   products across **all** linked child warehouses. To access, go to a product form and select the
   :guilabel:`On Hand` smart button.

   .. image:: stock_warehouses/on-hand.png
      :align: center
      :alt: Display stock across all linked warehouses.

To change the virtual stock location's :guilabel:`Location Type` to :guilabel:`View`, navigate to
:menuselection:`Inventory --> Configuration --> Locations`.

Click the virtual warehouse's stock location (e.g. `VWH/Stock`) that was :ref:`previously created
<inventory/routes/virtual-wh>`, from the :guilabel:`Locations` list.

On the location form, the under :guilabel:`Additional Information` heading, set the
:guilabel:`Location Type` to :guilabel:`View`. :guilabel:`Save` the changes.

.. image:: stock_warehouses/set-location-type-view.png
   :align: center
   :alt: Warehouse location types in location creation screen.

Example: Sell products from a virtual warehouse
===============================================

To sell products from multiple warehouses using a virtual parent location, the database must have at

> least **two** products and **two** warehouses configured — with at least **one** product with

Author: nice catch-- the "two products" part wasn't supposed to be there

   :align: center
   :alt: Delivery order with matching source and child locations.

.. important::
   The :guilabel:`Source Location` on the warehouse delivery form and the :guilabel:`Warehouse`
   under the :guilabel:`Other Info` tab on the sales order **must** match for products in the sales
   order to be pulled from different warehouses.

  - If the virtual warehouse is not in the :guilabel:`Source Location` field on the warehouse
    delivery form, retry product reservation by:

    - Running the scheduler: turn on :ref:`developer mode <developer-mode>`, and then go to
      :menuselection:`Inventory app --> Operations --> Run Scheduler`
    - Clicking :guilabel:`Check Availability` in the delivery order.
  - If the virtual warehouse is **not** assigned to the :guilabel:`Warehouse` field on the sales
    order, then cancel it and create a new sales order with the virtual warehouse set in the
    :guilabel:`Warehouse` field.
  - If the :guilabel:`Warehouse` field is missing on the sales order form, then the multiple
    children warehouses may not have been set up correctly. Review the :ref:`previous section
    <inventory/routes/child-wh>` to ensure the correct settings.

> .. danger::

Reviewer 2: The danger section makes sense to me, but I think we could provide a bit more guidance to users who have two or three-step delivery enabled:
-What is the effect of having a virtual address set as the output/packing location?
-In what situations would a user still want to use this workflow?
-Is there any workaround that makes this flow work normally for two and three-step users?
-Etc.

Also, since this warning seems to be very important, should we put it at the top of the doc? I don't think we want a two or three-step user to get to the end of the workflow before realizing that they can't use this fulfillment method.
